### PR TITLE
[FIXED JENKINS-47438] Keep track of recent approvals to prevent race conditions

### PR DESCRIPTION
**Pull request by stephenc**
[JENKINS-47438](https://issues.jenkins-ci.org/browse/JENKINS-47438)

@reviewbybees

### CODE REVIEW

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/ThrottleComputationQueueTaskDispatcher.java [hunk 1]
```diff
@@ -92,4 +134,14 @@ public int indexingCount(@CheckForNull Node node) {
         return result;
     }
 
+    private static class NonBlockedDetails {
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/ThrottleComputationQueueTaskDispatcher.java]: Maybe expressed more easily using a `WeakHashMap<Queue.Item, Long>`?

Comment (stephenc) [reply src/main/java/com/cloudbees/hudson/plugins/folder/computed/ThrottleComputationQueueTaskDispatcher.java]: Actually I think there's an even better way to express this

**Comment by stephenc** [comment]
@jglick a customer was facing a critical outage yesterday, test to follow

File: src/main/java/com/cloudbees/hudson/plugins/folder/computed/ThrottleComputationQueueTaskDispatcher.java [hunk 2]
```diff
@@ -92,4 +172,20 @@ public int indexingCount(@CheckForNull Node node) {
         return result;
     }
 
+    /**
+     * Extend {@link WeakReference} to allow association of secondary data without requiring excessive boxing of
+     * {@link Long}s
+     */
+    private static class TimestamppedWeakReference extends WeakReference<Queue.Item> {
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/computed/ThrottleComputationQueueTaskDispatcher.java]: typo

**Pull request merged**